pkg/operator/ceph/cluster/controller_test.go

```diff
@@ -27,116 +29,247 @@ import (
+			RequestCancelOrchestration: abool.New(),
+		}
+
+		// set up ClusterController
+		volumeAttachmentController := &attachment.MockAttachment{
+			MockList: func(namespace string) (*rookalpha.VolumeList, error) {
+				t.Log("test vol attach list")
+				return &rookalpha.VolumeList{Items: []rookalpha.Volume{}}, nil
+			},
+		}
+		operatorConfigCallbacks := []func() error{
+			func() error {
+				t.Log("test op config callback")
+				return nil
+			},
+		}
+		addCallbacks := []func() error{
+			func() error {
+				t.Log("test success callback")
+				return nil
+			},
+		}
+
+		// create the cluster controller and tell it that the cluster has been deleted
+		controller := NewClusterController(clusterdCtx, "", volumeAttachmentController, operatorConfigCallbacks, addCallbacks)
+		fakeRecorder := record.NewFakeRecorder(5)
+		controller.recorder = k8sutil.NewEventReporter(fakeRecorder)
+
+		// Create a fake client to mock API calls
+		// Make sure it has the fake CephCluster that is to be deleted in it
+		client := clientfake.NewClientBuilder().WithScheme(scheme).WithRuntimeObjects(fakeCluster.DeepCopy()).Build()
+
+		// Create a ReconcileCephClient object with the scheme and fake client.
+		reconcileCephCluster := &ReconcileCephCluster{
+			client:            client,
+			scheme:            scheme,
+			context:           clusterdCtx,
+			clusterController: controller,
+		}
+
+		req := reconcile.Request{NamespacedName: nsName}
+
+		resp, err := reconcileCephCluster.Reconcile(ctx, req)
+		assert.NoError(t, err)
+		assert.NotZero(t, resp.RequeueAfter)
+		event := <-fakeRecorder.Events
+		assert.Contains(t, event, "CephBlockPools")
+		assert.Contains(t, event, "my-block-pool")
+
+		blockedCluster := &cephv1.CephCluster{}
+		err = client.Get(ctx, nsName, blockedCluster)
+		assert.NoError(t, err)
+		status := blockedCluster.Status
+		assert.Equal(t, cephv1.ConditionDeleting, status.Phase)
+		assert.Equal(t, cephv1.ClusterState(cephv1.ConditionDeleting), status.State)
+		assert.Equal(t, corev1.ConditionTrue, cephv1.FindStatusCondition(status.Conditions, cephv1.ConditionDeleting).Status)
+		assert.Equal(t, corev1.ConditionTrue, cephv1.FindStatusCondition(status.Conditions, cephv1.ConditionDeletionIsBlocked).Status)
+
+		// delete blocking dependency
+		gvr := cephv1.SchemeGroupVersion.WithResource("cephblockpools")
+		err = clusterdCtx.DynamicClientset.Resource(gvr).Namespace(cephNs).Delete(ctx, "my-block-pool", metav1.DeleteOptions{})
+		assert.NoError(t, err)
+
+		resp, err = reconcileCephCluster.Reconcile(ctx, req)
+		assert.NoError(t, err)
+		assert.True(t, resp.IsZero())
+		event = <-fakeRecorder.Events
+		assert.Contains(t, event, "ClusterDeleting")
 
-	// The test does not return volumes on the second call
-	assert.NoError(t, controller.checkIfVolumesExist(clusterToDelete))
+		unblockedCluster := &cephv1.CephCluster{}
+		err = client.Get(ctx, nsName, unblockedCluster)
+		assert.NoError(t, err)
```

Comment:
I guess because of the deletion timestamp and once dependents are gone the deletion proceeds and then you hit IsNotFound. You cannot possibly catch the status update I believe, the timing is too short?. So L155 to L160 will never work.

I believe this diff would be sufficient:

```
diff --git a/pkg/operator/ceph/cluster/controller_test.go b/pkg/operator/ceph/cluster/controller_test.go
index bb63dac7f..09bfd20d6 100644
--- a/pkg/operator/ceph/cluster/controller_test.go
+++ b/pkg/operator/ceph/cluster/controller_test.go
@@ -31,6 +31,7 @@ import (
 	"github.com/stretchr/testify/assert"
 	"github.com/tevino/abool"
 	corev1 "k8s.io/api/core/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/apimachinery/pkg/types"
@@ -150,7 +151,8 @@ func TestReconcile_DeleteCephCluster(t *testing.T) {
 
 		unblockedCluster := &cephv1.CephCluster{}
 		err = client.Get(ctx, nsName, unblockedCluster)
-		assert.NoError(t, err)
+		assert.Error(t, err)
+		assert.True(t, true, kerrors.IsNotFound(err))
 		assert.Empty(t, unblockedCluster.Finalizers)
 		status = unblockedCluster.Status
 		assert.Equal(t, cephv1.ConditionDeleting, status.Phase)
```

Hopefully, I did not miss anything.